contracts/ts/__tests__/E2E.test.ts

```diff
@@ -0,0 +1,690 @@
+                salt: circuitInputs.salt,
+                stateTreeRoot: circuitInputs.stateTreeRoot,
+                deactivatedKeysRoot: circuitInputs.deactivatedKeysRoot,
+                stateTreeInclusionProof: circuitInputs.stateTreeInclusionProof,
+                oldCreditBalance: circuitInputs.oldCreditBalance,
+                newCreditBalance: circuitInputs.newCreditBalance,
+                stateLeafTimestamp: circuitInputs.stateLeafTimestamp,
+                deactivatedKeysInclusionProof: circuitInputs.deactivatedKeysInclusionProof,
+                deactivatedKeyIndex: circuitInputs.deactivatedKeyIndex,
+                c1: circuitInputs.c1,
+                c2: circuitInputs.c2,
+                coordinatorPubKey: circuitInputs.coordinatorPubKey,
+                encPrivKey: circuitInputs.encPrivKey,
+                c1r: circuitInputs.c1r,
+                c2r: circuitInputs.c2r,
+                z: circuitInputs.z,
+                nullifier: circuitInputs.nullifier,
+                pollId: circuitInputs.pollId
+            },
+            gnkWasmFile,
+            gnkZkeyFile
+        )
+
+        // Check proof validity.
+        const isValidProof = await groth16.verify(
+            rawGnkVk,
+            publicSignals,
+            proof
+        )
+
+        // Format proof for verifier contract.
+        const formattedProof = formatProofForVerifierContract(proof);
+
+        // Generate new key from deactivated key.
+        const tx = await messageProcessorContract.generateNewKeyFromDeactivated(
+            message.asContractParam(),
+            coordinatorKeyPair.pubKey.asContractParam(),
+            encPubKey.asContractParam(),
+            pollContract.address,
+            formattedProof
+        )
+        const receipt = await tx.wait()
+
+        // State update.
+        numOfGeneratedKeys += 1
+
+        // Events.
+        const logAttemptKeyGeneration = pollContract.interface.parseLog(receipt.logs[0])
+
+        // Checks.
+        expect(isValidProof).toBe(true)
+        expect(logAttemptKeyGeneration.name).toBe("AttemptKeyGeneration")
+        expect(logAttemptKeyGeneration.args._message.data.map((x: any) => BigInt(x))).toStrictEqual(message.data)
+        expect(String(logAttemptKeyGeneration.args._encPubKey.x)).toBe(String(encPubKey.asContractParam().x))
+        expect(String(logAttemptKeyGeneration.args._encPubKey.y)).toBe(String(encPubKey.asContractParam().y))
+        expect(Number(logAttemptKeyGeneration.args._newStateIndex)).toBe(Number(circuitInputs.numSignUps) + Number(numOfGeneratedKeys))
+    })
+
+    it("should publish the vote message for the user", async () => {
+        // Prepare data.
+        const messageNonce = BigInt(1)
+        const voteOptionIndex = 0
+        const newVoteWeight = 1
+
+        // Prepare the command.
+        const command: PCommand = new PCommand(
+            BigInt(userStateIndex.toString()),
+            user1KeyPair.pubKey,
```

Comment:
The user should be voting with the newUser1KeyPair.pubKey here, right?

Reply:
Correct, great catch!

Reply:
https://github.com/0x3327/maci/pull/33/commits/e9a120a4ba8be54520281719c0df28f9ff38b4a1 done